teps/0123-specifying-on-demand-retry-in-pipelinetask.md

```diff
@@ -0,0 +1,183 @@
+creation-date: '2022-09-16'
+last-updated: '2022-10-11'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0123: Specifying on-demand-retry in a PipelineTask
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+  - [Use Cases](#use-cases)
+    - [CD Use Case](#cd-use-case)
+    - [CI Use Case](#ci-use-case)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a mechanism to allow users to choose when to retry a failed `pipelineTask` of a pipeline.
+This kind of `on-demand` retry of a `pipelineTask` is executed with the original `params` and `specifications`.
+The `on-demand` retry of a `pipelineTask` allow users to continue to retry until the task succeeds. After the offending task
+succeeds, the rest of the `pipeline` continues executing as usual until the completion. This retry of a `pipelineTask`
+just like the existing `retry` strategy is only applicable within the context of an original `pipelineRun` i.e.
+a `pipelineTask` can only be retried until the `pipelineRun` that it's part of is `running`.
+
+## Motivation
+
+The primary motivation of this proposal is to enable pipeline authors to overcome any flakes without re-running
+the whole pipeline but to re-run the flakes and continue the execution until the pipeline finishes. The main motivation
+of this proposal is to enable pipeline authors to specify **when** to re-run such flakes.
+
+Tekton controller does support [retries](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#using-the-retries-field)
+in which a pipeline author can specify the number of the times a task can be retried when it fails. The `retries` are designed
+such that once a task fails, it is retried immediately. Other CI/CD systems supports an additional retry strategies, such as:
+
+* Argo Workflow offers a variety of retry policies. One of them is [Back-Off](https://argoproj.github.io/argo-workflows/retries/#back-off),
+in which a user can configure the delay between retries.
+
+* GitHub action supports a retry of an action allowing to set some delay between attempts through [attempt_delay](https://github.com/marketplace/actions/retry-action#attempt_delay).
+
+These additional retry strategies allows users to specify a reasonable delay during which a transient failure can be
+fixed.
+
+Jenkins takes `retry` one step further where a user can specify [conditions](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times)
+for which a task/stage must be retried. For example, when a stage fails because of an underlying infrastructure issue
+rather than the execution of the `script`. Jenkins stage author can specify a condition `agent` which will allow
+retry once the connection to an agent is fixed.
+
+Jenkins declarative pipeline supports restarting stages with the same parameters and actions as documented in
```

Comment:
This TEP **does not** consider the partial execution of a pipeline that reached completion, whether successful or not. The key assumption for the retries in this TEP is that the pipeline is **still running**. 

If a `PipelineTask` marked for "on-demand-retry" fails, the execution of the `Pipeline` continues until either of the following happens:

- the `PipelineTask` is successful, and the associated branch of execution is resumed. This may happen if a retry is requested and is successful. 
- timeout is reached
- the `PipelineRun` is cancelled

Reply:
Reading through this section again it seems like other CI systems provide a way to configure a delay before configuring other retries. Jenkins allows users to specify conditional logic before retrying. Jenkins Declarative pipelines does support restarting stages on demand but like you pointed out that does not keep the current pipeline running.
So, I think the current proposal is a bit unique in that respect. I don't think this by itself should mean should should not pursue on demand retries just that we should consider the alternative solutions for the use cases as well before implementing this

Reply:
The clear distinction here is **You can restart any completed Declarative Pipeline**, this proposal is not applicable if the pipeline is completed.

Reply:
Like @afrittoli explained this proposal is only applicable until one of those three criteria is met.

Reply:
thanks @dibyom, please elaborate on this

> other CI systems provide a way to configure a delay before configuring other retries.

Delay is applicable in terms of retries, if a stage fails and asked to retry, a delay can be introduced to schedule a stage for the execution after that interval. 

> Jenkins allows users to specify conditional logic before retrying. 

Conditional logic is part of retry. Jenkins support `try-catch` in addition to `retry` which is much more robust when used along with `retry`.

https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#catcherror-catch-error-and-set-build-result-to-failure

>  I don't think this by itself should mean should should not pursue on demand retries just that we should consider the alternative solutions for the use cases as well before implementing this

What are the alternative solutions here apart from custom task? This is a very common use case and if we decide to support it, it must live as a first class solution rather than a custom task.